### PR TITLE
[terra-graphs] custom tick values overflow past axis limits

### PR DESCRIPTION
**Pull request by udaymattam**
### Summary

**What was changed:**
added filter method to remove the grid lines outside the grid.

**Why it was changed:**
The grid lines overflow outside the grid, especially when the browser size is reduced.

### Testing

This change was tested us
before fix
screenshot-2

after fix
Screenshot 2023-06-06 at 8 13 33 PM

- [ ] WDIO
- [ ] Jest
- [X] Visual testing (please attach a screenshot or recording)
- [ ] Other (please describe below)
- [ ] No tests are needed

### Reviews

In addition to engineering reviews, this PR needs:

- [ ] UX review
- [ ] Accessibility review
- [ ] Functional review

### Additional Details

**This PR resolves:**

UXPLATFORM-XXXX 

---

Thank you for contributing to Terra.
@cerner/terra

### CODE REVIEW

File: packages/carbon-graphs/CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,9 @@
 
 ## Unreleased
 
+* Added
+  * Added filter method for overflow tick values.
+
```
Comment (sdadn) [inline packages/carbon-graphs/CHANGELOG.md:7]:
```suggestion
* Fixed
  * Fixed tick lines overflowing outside the grid container when using custom tick values.

```

File: packages/terra-graphs-docs/src/example-datasets/graphConfigObjects/General/generalDefaultXAxisOverflow.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+          180,
+          186,
+          192,
+          198,
+          204,
+          210,
+          216,
+          222,
+        ],
+        // ].filter(value => value >= '40' && value <= '168'),
```
Comment (sdadn) [inline packages/terra-graphs-docs/src/example-datasets/graphConfigObjects/General/generalDefaultXAxisOverflow.js:46]: Do we still need this?

**Review by sdadn — CHANGES_REQUESTED** [review]

File: packages/carbon-graphs/CHANGELOG.md (hunk 1 above)
Comment (udaymattam) [reply packages/carbon-graphs/CHANGELOG.md]: Updated

File: packages/carbon-graphs/src/js/controls/Graph/helpers/helpers.js [hunk 3]
```diff
@@ -167,10 +167,17 @@ const translateVerticalGrid = (axis, config) => {
     : constants.TICK_ORIENTATION.BOTTOM;
   if (utils.notEmpty(config.axis.x.ticks.values)) {
     const ticks = config.axis.x.ticks.values;
-    xAxisGrid = axis.x
-      .tickValues(processTickValues(ticks))
-      .tickSize(getYAxisHeight(config) * tickSizeMultiplicand, 0, 0)
-      .tickFormat('');
+    if (config.axis.x.domain === undefined) {
```
Comment (sdadn) [inline packages/carbon-graphs/src/js/controls/Graph/helpers/helpers.js:170]: We can use the `utils.isUndefined` method 

https://github.com/cerner/terra-graphs/blob/5688742718cc2e135e549e233d729f6c739ee1f9/packages/carbon-graphs/src/js/helpers/utils.js#L48-L55

```suggestion
    if (utils.isUndefined(config.axis.x.domain)) {
```

File: packages/carbon-graphs/src/js/helpers/axis.js [hunk 4]
```diff
@@ -127,17 +127,25 @@ const prepareXAxis = (
...
   if (isXAxisOrientationTop(orientation)) {
     d3Axis = d3.axisTop(scale);
   }
-  d3Axis
-    .ticks(
-      Math.max(width / constants.MAX_TICK_VARIANCE, constants.MIN_TICKS),
-    )
-    .tickValues(processTickValues(tickValues))
-    .tickFormat(format);
+  if (tickValues === undefined) {
```
Comment (sdadn) [inline packages/carbon-graphs/src/js/helpers/axis.js:137]: same here (`utils.isUndefined`)

File: packages/carbon-graphs/src/js/helpers/axis.js [hunk 5]
```diff
@@ -127,17 +127,25 @@ const prepareXAxis = (
   width,
   format,
   orientation = AXES_ORIENTATION.X.BOTTOM,
+  lowerLimit = null,
+  upperLimit = null,
```
Comment (sdadn) [inline packages/carbon-graphs/src/js/helpers/axis.js:131]: Why are these being defaulted to null? This is effectively evaluating the value against 0 which may not be correct.

File: packages/carbon-graphs/tests/wdio/Graphs/BarGraph/__snapshots__/reference/terra-default-theme/en/chrome_large/bar-graph-spec/bar_dynamic_data_panning_left_1.png
Comment (sdadn) [inline packages/carbon-graphs/tests/wdio/Graphs/BarGraph/__snapshots__/reference/terra-default-theme/en/chrome_large/bar-graph-spec/bar_dynamic_data_panning_left_1.png:1]: These screenshot updates do not look correct; new axis limits are being added.

File: packages/carbon-graphs/src/js/helpers/axis.js [hunk 6]
```diff
@@ -108,6 +108,12 @@ const processTickValues = (ticks) => {
   return ticks.map((t) => (utils.isDate(t) ? utils.parseDateTime(t) : t));
 };
 
+const hasDateValues = (array) => {
+  if (utils.isDate(array)) {
+    return true;
+  }
+  return false;
+};
```
Comment (sdadn) [inline packages/carbon-graphs/src/js/helpers/axis.js:116]: What is the purpose of this function when `utils.isDate` already exists?

File: packages/carbon-graphs/tests/wdio/Graphs/GanttChart/__snapshots__/reference/terra-default-theme/en/chrome_tiny/gantt-chart-spec/gantt_datetime_bucket.png
Comment (sdadn) [inline packages/carbon-graphs/tests/wdio/Graphs/GanttChart/__snapshots__/reference/terra-default-theme/en/chrome_tiny/gantt-chart-spec/gantt_datetime_bucket.png:1]: this snapshot looks differerent

File: packages/carbon-graphs/src/js/helpers/axis.js [hunk 7]
```diff
@@ -127,17 +133,36 @@ const prepareXAxis = (
...
-    .ticks(
-      Math.max(width / constants.MAX_TICK_VARIANCE, constants.MIN_TICKS),
-    )
-    .tickValues(processTickValues(tickValues))
-    .tickFormat(format);
+  if (!tickValues || tickValues.length === 0) {
+    d3Axis
+      .ticks(
+        Math.max(width / constants.MAX_TICK_VARIANCE, constants.MIN_TICKS),
+      )
```
Comment (sdadn) [inline packages/carbon-graphs/src/js/helpers/axis.js:147]: Cane we make this a single line for better readability?
```suggestion
      .ticks(Math.max(width / constants.MAX_TICK_VARIANCE, constants.MIN_TICKS))
```

File: packages/carbon-graphs/src/js/helpers/axis.js [hunk 8]
```diff
@@ -127,17 +133,36 @@ const prepareXAxis = (
...
+      .tickValues(processTickValues(tickValues))
+      .tickFormat(format);
+  } else if (hasDateValues(tickValues)) {
+    const filteredTickValues = tickValues.filter((value) => {
+      const date = new Date(value);
+      return date >= lowerLimit && date <= upperLimit;
```
Comment (sdadn) [inline packages/carbon-graphs/src/js/helpers/axis.js:153]: Are date, lowerLimit and upperLimit the same type here? Can you do a console.log and post a screenshot of the results here?

File: packages/carbon-graphs/tests/wdio/Graphs/General/__snapshots__/reference/terra-default-theme/en/chrome_tiny/general-spec/datetime_bucket.png
Comment (sdadn) [inline packages/carbon-graphs/tests/wdio/Graphs/General/__snapshots__/reference/terra-default-theme/en/chrome_tiny/general-spec/datetime_bucket.png:1]: this snapshot also looks different

File: packages/carbon-graphs/src/js/helpers/axis.js (hunk 8 above)
Comment (udaymattam) [reply packages/carbon-graphs/src/js/helpers/axis.js:153]: yes, same type
Screenshot 2023-06-28 at 11 36 11 AM
Screenshot 2023-06-28 at 11 36 46 AM

File: packages/carbon-graphs/tests/unit/controls/Graph/GraphAxes-spec.js [hunk 9]
```diff
@@ -1212,6 +1214,70 @@ describe('Graph - Axes', () => {
...
+      };
+      graph = new Graph({ ...getAxes(localeAxisObj) });
+
+      const allXAxisElements = document.querySelectorAll(`.${styles.axisX}`);
+      const xTicks = [];
+      const expectedXTicks = [2, 3, 4, 5, 6, 7, 8];
+
+      for (let i = 1; i < allXAxisElements[0].childNodes.length; i += 1) {
+        xTicks.push(parseInt(allXAxisElements[0].childNodes[i].querySelector('text').textContent, 10));
+      }
```
Comment (sdadn) [inline packages/carbon-graphs/tests/unit/controls/Graph/GraphAxes-spec.js:1245]: We can use the [map function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) for an array:
```suggestion
const xTicks = allXAxisElements[0].childNodes.map((node)=>parseInt(node.querySelector('text').textContent,10));
```

File: packages/carbon-graphs/tests/unit/controls/Graph/GraphAxes-spec.js [hunk 10]
```diff
@@ -1212,6 +1214,70 @@ describe('Graph - Axes', () => {
       expect(yTicks).toEqual(expectedYTicks);
       expect(y2Ticks).toEqual(expectedY2Ticks);
     });
+
+    it('tesing filter method for numeric', () => {
```
Comment (sdadn) [inline packages/carbon-graphs/tests/unit/controls/Graph/GraphAxes-spec.js:1218]:
```suggestion
    it('filters out numeric axis values exceeding upper and lower limits', () => {
```

File: packages/carbon-graphs/tests/unit/controls/Graph/GraphAxes-spec.js [hunk 11]
```diff
@@ -1212,6 +1214,70 @@ describe('Graph - Axes', () => {
...
+      const expectedXTicks = [2, 3, 4, 5, 6, 7, 8];
+
+      for (let i = 1; i < allXAxisElements[0].childNodes.length; i += 1) {
+        xTicks.push(parseInt(allXAxisElements[0].childNodes[i].querySelector('text').textContent, 10));
+      }
+
+      expect(xTicks).toEqual(expectedXTicks);
+    });
+
+    it('tesing filter method for DateTicks', () => {
```
Comment (sdadn) [inline packages/carbon-graphs/tests/unit/controls/Graph/GraphAxes-spec.js:1250]:
```suggestion
    it('filters out datetime axis values exceeding upper and lower limits', () => {
```

File: packages/carbon-graphs/tests/unit/controls/Graph/GraphAxes-spec.js [hunk 12]
```diff
@@ -1212,6 +1214,70 @@ describe('Graph - Axes', () => {
...
+      const expectedXTicks = [
+        new Date(2016, 2).toISOString(),
+        new Date(2016, 3).toISOString(),
+        new Date(2016, 4).toISOString(),
+      ];
+      for (let i = 1; i < allXAxisElements[0].childNodes.length; i += 1) {
+        const { textContent } = allXAxisElements[0].childNodes[i].querySelector('text');
+        const dateTimeValue = new Date(textContent).toISOString();
+        xTicks.push(dateTimeValue);
+      }
```
Comment (sdadn) [inline packages/carbon-graphs/tests/unit/controls/Graph/GraphAxes-spec.js:1278]: Same usage of the map function here:

**Review by sdadn — CHANGES_REQUESTED** [review]

**Review by sycombs — APPROVED** [review]

**Review by sdadn — APPROVED** [review]

**Review by supreethmr — APPROVED** [review]

**Pull request merged**